File: src/main/java/ting/leetcode/java/arrays101/LeastNumberOfUniqueIntegersAfterKRemovals.java
package ting.leetcode.java.arrays101;

import java.util.Arrays;
import java.util.PriorityQueue;

public class LeastNumberOfUniqueIntegersAfterKRemovals {
    public int findLeastNumOfUniqueInts(int[] arr, int k) {
        Arrays.sort(arr);
        PriorityQueue<Integer> elements = new PriorityQueue<>();
        int count = 0;
        int current = arr[0];
        for (int num : arr) {
            if (current != num) {
                elements.add(count);
                current = num;
                count = 1;
            } else {
                count++;
            }
        }

        elements.add(count);


        int accumulator = 0;
        while (accumulator < k) {
            accumulator += elements.poll();
        }

        if (accumulator == k)
            return elements.size();
        return elements.size() + 1;
    }
}

File: src/main/kotlin/ting/leetcode/arrays101/LeastNumberOfUniqueIntegersAfterKRemovals.kt
package ting.leetcode.arrays101

import java.util.*

class LeastNumberOfUniqueIntegersAfterKRemovals {
    fun findLeastNumOfUniqueInts(arr: IntArray, k: Int): Int {
        arr.sort()
        val queue = PriorityQueue<Int>()
        var count = 0
        var current = arr.first()

        for (num in arr) {
            if (current != num) {
                queue.add(count)
                current = num
                count = 1
            } else count++
        }

        queue.add(count)

        var accumulator = 0
        while (accumulator < k)
            accumulator += queue.poll()

        return if (accumulator == k) queue.size else queue.size + 1
    }
}
